### PR TITLE
Task/ctv 2663 html5 reference app with js ima

### PR DESCRIPTION
**Pull request by rblaa**
### JIRA
https://truextech.atlassian.net/browse/CTV-2663

### Description
Migrate ctv ref app to use Google IMA for main video and stiched in ads.

### Testing
* Run locally via npm start, restart video, play/pause/seek around ads, interact with ads, complete them, exit them early, etc.

### Commit Message Subject(s)
CTV-2663: HTML5 Reference App w/ JS IMA

### Additional Context
n/a

### Related PRs
n/a

### Checks
- [ ] Includes unit test coverage _(if applicable)_
- [x] Includes functional test coverage _(at feature-level, if applicable)_
- [x] Proposed commit messages follow [team conventions](https://github.com/socialvibe/adlabs-wiki/blob/develop/git/README.md#commits)

### CODE REVIEW

File: src/components/video-controller.js [hunk 1]
```diff
@@ -319,27 +419,54 @@ export class VideoController {
...
         this.hideControlBar();
 
         this.stopVideo(); // avoid multiple videos, e.g. for platforms like the PS4
 
-        // ensure main video is logically at the ad start for when it resumes
-        this.initialVideoTime = adBreak.startTime;
+        // ensure main video is logically at the fallback videos for when it resumes
+        // We just need to skip over the placeholder video of this interactive ad wrapper.
+        adBreak.placeHolderDuration = googleAd.getDuration();
+        this.initialVideoTime = adBreak.fallbackStartTime;
```
Comment (kylelamx) [inline src/components/video-controller.js:467]: how about.. linearAdStartTime?

File: src/components/video-controller.js [hunk 2]
```diff
@@ -379,36 +517,31 @@ export class VideoController {
                     return;
                 }
             } else if (!adBreak.started) {
-                this.startAd(adBreak);
+                // We will get ad events when the ad is encountered
```
Comment (kylelamx) [inline src/components/video-controller.js:520]: how about we add "...ad events _from Google IMA stream event_..."?

**Review by kylelamx — APPROVED** [review]
Looking Great!

File: src/data/video-streams.json [hunk 3]
```diff
@@ -4,20 +4,7 @@
...
-      },
-      {
-        "breakId": "midroll-1",
-        "timeOffset": "00:08:05.000",
-        "videoAdDuration": "92",
-        "vastUrl": "https://qa-get.truex.com/22105de992284775a56f28ca6dac16c667e73cd0/vast/config?dimension_1=sample-video&dimension_2=1&dimension_3=sample-video&dimension_4=5678&dimension_5=evergreen&stream_position=midroll&stream_id=5678"
-      }
-    ]
+    "google_content_id" : "2496857",
+    "google_video_id" : "truex-content22-4k"
```
[the remark this answers is not in the document]
Comment (rblaa) [reply src/data/video-streams.json:8]: I find things more maintainable doing it within the repo.
E.g. I cannot update that location, and having it in the project shows the data format to all devs immediately.

File: src/components/ad-break.js [hunk 4]
```diff
@@ -1,38 +1,29 @@
 /**
- * Describes a single ad break that maps to 1 or more fallback ad videos in the main video
- * (ads are assumed to be sitched in), that furthermore describes a true[X] interactive ad to show
- * over top of the main video when the ad break is encountered during playback.
+ * Describes a single timespan within a full video that contains consecutive stitched-in ad videos.
+ * The first video in each ad break is assumed to be the placeholder video for the true[X] interactive ad, and
```
[the remark this answers is not in the document]
Comment (rblaa) [reply src/components/ad-break.js:3]: Right. The code does handle that case. 
I will correct to say "if a truex ad..."

File: src/components/video-controller.js [hunk 5]
```diff
@@ -319,27 +419,54 @@ export class VideoController {
...
+    startAd(googleAd) {
+        const podInfo = googleAd.getAdPodInfo();
+        const adBreak = this.adBreaks[podInfo.getPodIndex()];
+        if (!adBreak) return;
+        if (googleAd.getAdSystem() != 'trueX') return; // ignore non-trueX ads
+
+        // For true[X] IMA integration, the first ad in an ad break points to the interactive ad,
+        // the remaining ones are the fallback ad videos.
+        const adPosition = podInfo.getAdPosition();
+        if (adPosition != 1) return;
```
[the remark this answers is not in the document]
Comment (rblaa) [reply src/components/video-controller.js:440]: I have no other way to recognize things then, unless we look at the description containing truex.com

File: src/components/video-controller.js (hunk 1 above)
Comment (rblaa) [reply src/components/video-controller.js:467]: well, I think of the remaining videos as the fallback ad videos, to use when the interactive ad cannot be, or is abandoned.

File: src/components/video-controller.js (hunk 2 above)
Comment (rblaa) [reply src/components/video-controller.js:520]: ok, will fix

File: src/components/video-controller.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rblaa) [reply src/components/video-controller.js:440]: I tweaked the assumptions to be more explicit.

**Pull request merged**